### DISCUSSION TITLE
Some Ideas

### DISCUSSION THREAD

**Discussion by ChiBear** [kimlimjustin/xplorer#79] [Ideas]
Here are some ideas for potential consideration:

---

**1.**  Since this file explorer is built sorta like a browser, I think some browser-like functionality could be implemented, such as:
 
a. Middle mouse button on a folder to open it in a new tab. I use this a LOT in browser, so this functionality in xplorer would be amazing to have

b. The ability to pin a tab to the top. It would be complementary to the already existing sidebar, with its use case potentially for more specific folders that would clutter the sidebar. 

c. A dropdown at the end of the tab area that'll show all tabs, for when too many tabs have been opened

d. Leaving the 'add tab' icon at the end, separate from the tabs themselves. Having to scroll to the end just to open a tab is kinda annoying tbh :/

**2.**  Some potential styling/ UI ideas:

a. Allow changing of name of individual tabs

b. Allow changing of color of individual tabs

c. Allow grouping of tabs 

d. Ability to close the sidebar (this could be complementary to the other idea I saw about making this app a file reader of sorts as it gives more screen real estate for that)

---

Some of these ideas I legitimately want to see implemented (like the middle mouse button), and others are just me spitballing (like the grouping of tabs).

If I get any more ideas, I'll add it here or create a new discussion. 

I like spitballing ideas, whether you decide to take my ideas seriously or not is up to you. 

Have a good day :)

**Comment by kimlimjustin** [comment]
Thanks for the ideas, I'll gradually implement them (can't be so soon since I got a lot of other things to do too but I'll try my best). TYSM

**Reply by ChiBear** [reply]
No rush, man. Do your thing, I'll be here if and when they do get implemented.

**Reply by kimlimjustin** [reply]
Okay thanks!

**Comment by vitalics** [comment]
+1 idea - custom context menu actions. As for me - I create integration between finder and vscode and now I can open the current folder with vscode.
I do this thing via the Automator app.  Maybe it will be great to add such options as custom actions. See attached the screenshot.

![image](https://user-images.githubusercontent.com/8816260/143187780-6abfad9f-097b-45cc-8309-d1e8811ca43c.png)

Also in windows - I was using tortoisegit as a git client. It integrates with windows explorer.

**Reply by kimlimjustin** [reply]
Do you mean the ability from the user side to add a custom context menu command or the ability to inherit the context menu from the default file explorer/finder into Xplorer?

If you mean the first, I think that creating an "extension" support would be a better option just like vscode does, I'm planning on implementing it and indeed I created an easy to change context-menu's menu configuration, you can see it [here](https://github.com/kimlimjustin/xplorer/tree/dd0d8146f1d93dc313ae5c56af52bf2bc8149658/src/Components/ContextMenu/configs). but got little problems about the security concern, soo yeah, might take some time to get the extension support to be implemented, but of course, will be implemented before the 1.0 version is out. 

If you mean the second, I meant to implement that too, but can't really find any documentation online of the context menu's API in Rust. so maybe that'll be done by calling c from Rust, but yeah, it'll take so much works cuz I need to take care of three platforms, from Windows, Linux to macOS (and my macOS virtual machine for some reasons is broken), so I decided to not touch it for now, but probably will in the future.

Also, for now, there're only two main developers (who both are still high school students) working on Xplorer now, so can't really make it really fast, so yeah if you want to help, it's will be much appreciated.
